[redis] fix pipeline return types

chdsbd opened the pull request:
`Pipeline` is a subclass of `Redis`, but instead of the commands returning results, they return `Pipeline`.

Previously the types said `Redis().pipeline().rpush("mylist")` returned `int`. The updated types now type the response as `Pipeline`.

With this change the following code now type checks.
```python
from redis import Redis
r = Redis()
list_length, _ = r.pipeline().rpush("mylist").ltrim("mylist", 0, 100).execute()
```

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
-    def reset(self): ...
+    def reset(self) -> None: ...
     def multi(self) -> None: ...
-    def execute_command(self, *args, **kwargs): ...
```
`execute_command`, along with `parse_response`, `watch`, and `unwatch` were duplicated in the methods I copied from `Redis`, so I deleted these and updated the copied ones below.

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
-    def unwatch(self): ...
     def script_load_for_pipeline(self, script): ...
-
-class StrictPipeline(BasePipeline, StrictRedis): ...
```
`StrictPipeline` and `BasePipeline` don't exist in the library, so I think it's safe to use use the name `Pipeline`: https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/redis/client.py

srittau replied:
This probably was leftover from an old redis-py version.

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
-
-class StrictPipeline(BasePipeline, StrictRedis): ...
-class Pipeline(BasePipeline, Redis): ...
+    # methods from `Redis` client.
+    # arguments match `Redis`, but with some exceptions return types are
+    # `Pipeline` or `None`
```
Basically all the commands have the same arguments as `Redis`, but the return types are different.

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+    def rpoplpush(self, src, dst) -> Pipeline: ...
+    def rpush(self, name: _Value, *values: _Value) -> Pipeline: ...
+    def rpushx(self, name, value) -> Pipeline: ...
+    def sort(
+        self,
+        name: _Key,
+        start: Optional[int] = ...,
+        num: Optional[int] = ...,
+        by: Optional[_Key] = ...,
+        get: Optional[Union[_Key, Sequence[_Key]]] = ...,
+        desc: bool = ...,
+        alpha: bool = ...,
+        store: Optional[_Key] = ...,
+        groups: bool = ...,
+    ) -> Pipeline: ...
```
`Redis().sort()` is overloaded to provide an accurate return type, but for `Pipeline` the return type is always `Pipeline`, so I removed the overloads.

https://github.com/python/typeshed/blob/e2fd852952ff807c60819f8840e8b23664cc7593/stubs/redis/redis/client.pyi#L495-L534

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+    def smembers(self, name: _Key) -> Pipeline: ...
+    def smove(self, src, dst, value) -> Pipeline: ...
+    def spop(self, name, count: Optional[int] = ...) -> Pipeline: ...
+    def srandmember(self, name: _Key, number: Optional[int] = ...) -> Pipeline: ...
```
I removed the overload here too for the same reason.

https://github.com/python/typeshed/blob/e2fd852952ff807c60819f8840e8b23664cc7593/stubs/redis/redis/client.pyi#L555-L558

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+        groups: bool = ...,
+    ) -> Pipeline: ...
+    def scan(self, cursor: int = ..., match: Optional[_Key] = ..., count: Optional[int] = ...) -> Pipeline: ...
+    def scan_iter(self, match: Optional[Text] = ..., count: Optional[int] = ...) -> Iterator[Any]: ...
```
The `scan_iter` and other `_iter` methods return generators that explode at runtime.

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+    def script_flush(self) -> Pipeline: ...
+    def script_kill(self) -> Pipeline: ...
+    def script_load(self, script) -> Pipeline: ...
+    def register_script(self, script: Union[Text, _StrType]) -> Script: ...
```
unlike other methods, this one actually returns `Script`

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+    def pubsub_numpat(self) -> Pipeline: ...
+    def monitor(self) -> Monitor: ...
+    def cluster(self, cluster_arg: str, *args: Any) -> Pipeline: ...
+    def client(self) -> Any: ...
```
this method explodes when called

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+    def pubsub_channels(self, pattern: _Key = ...) -> Pipeline: ...
+    def pubsub_numsub(self, *args: _Key) -> Pipeline: ...
+    def pubsub_numpat(self) -> Pipeline: ...
+    def monitor(self) -> Monitor: ...
```
This also returns `Monitor` instead of `Pipeline`

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+        xx: bool = ...,
+        keepttl: bool = ...,
+    ) -> Pipeline: ...
+    def __setitem__(self, name, value) -> None: ...
```
should all the dunder methods be organized at the top?

srittau replied:
Personally, I try to use the order the implementation uses. But it is your choice.

chdsbd commented on stubs/redis/redis/client.pyi:
```diff
@@ -684,38 +684,297 @@ class PubSub:
+    def object(self, infotype, key) -> Pipeline: ...
+    def ping(self) -> Pipeline: ...
+    def save(self) -> Pipeline: ...
+    def sentinel(self, *args) -> None: ...
```
this deprecated method doesn't do or return anything: https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/redis/client.py#L1448-L1451

srittau reviewed (changes requested):
Thank you for this PR and sorry for the long delay. Only one issue below.

chdsbd commented:
I updated `Pipeline` to inherit from Redis and added `# type: ignore [override]` to every method that returns `Pipeline`. Not all of these ignore statements are necessary because some methods in parent class are missing return types.

The ordering of methods should match the `redis-py` implementation. The `Pipeline` defines a few methods but inherits most methods from the `Redis` client.

Akuli commented:
I think `# type: ignore[override]` is a non-standard mypy thing? This PR however uses `# type: ignore [override]`, with a space. Maybe type checkers understand that to be same as just `# type: ignore`?
